share/functions/umask.fish

```diff
@@ -1,117 +1,124 @@
+    end
 
-            switch $i
-                case '=*'
-                    set mode set
-                    set i (echo $i| cut -c 2-)
-
-                case '+*'
-                    set mode add
-                    set i (echo $i| cut -c 2-)
-
-                case '-*'
-                    set mode remove
-                    set i (echo $i| cut -c 2-)
-
-                case '*'
-                    if not count $implicit_all >/dev/null
-                        printf (_ "%s: Invalid mask '%s'\n") umask $argv >&2
-                        return
-                    end
-                    set mode set
-            end
+    # Test if argument is a valid symbolic mask. Note that the basic pattern allows one illegal
+    # pattern: who and perms without a mode such as "urw". We test for that below after splitting
+    # the rights using the pattern and testing for that invalid combination.
+    set -l basic_pattern '([ugoa]*)([=+-]?)([rwx]*)'
+    if not string match -qr "^$basic_pattern(,$basic_pattern)*\$" -- $argv
+        printf (_ "%s: Invalid mask '%s'\n") umask $argv >&2
+        return 1
+    end
 
-            if not echo $perm | __fish_sgrep -E '^(r|w|x)*$' >/dev/null
-                printf (_ "%s: Invalid mask '%s'\n") umask $argv >&2
-                return
-            end
+    # Insert inverted umask into res variable.
+    set -l tmp (string split '' $umask)
+    set -l res
+    set res[1] (math 7 - $tmp[2])
+    set res[2] (math 7 - $tmp[3])
+    set res[3] (math 7 - $tmp[4])
+
+    for rights in (string split , $argv)
+        set -l match (string match -r "^$basic_pattern\$" $rights)
+        set -l scope $match[2]
+        set -l mode $match[3]
+        set -l perms $match[4]
+        if test -n "$scope" -a -z "$mode"
+            printf (_ "%s: Invalid mask '%s'\n") umask $argv >&2
+            return 1
+        end
+        if test -z "$scope"
+            set scope a
+        end
+        if test -z "$mode"
+            set mode =
+        end
 
-            set val 0
-            if echo $i | __fish_sgrep 'r' >/dev/null
-                set val 4
-            end
-            if echo $i | __fish_sgrep 'w' >/dev/null
-                set val (math $val + 2)
-            end
-            if echo $i | __fish_sgrep 'x' >/dev/null
-                set val (math $val + 1)
-            end
+        set -l scopes_to_modify
+        string match -q '*u*' $scope
+        and set scopes_to_modify 1
+        string match -q '*g*' $scope
+        and set scopes_to_modify $scopes_to_modify 2
+        string match -q '*o*' $scope
+        and set scopes_to_modify $scopes_to_modify 3
+        string match -q '*a*' $scope
+        and set scopes_to_modify 1 2 3
+
+        set -l val 0
+        if string match -q '*r*' $perms
+            set val 4
+        end
+        if string match -q '*w*' $perms
+            set val (math $val + 2)
+        end
+        if string match -q '*x*' $perms
+            set val (math $val + 1)
```

Comment:
True but that will save at most one math invocation at the cost of code that isn't as clear.